### PR TITLE
Safe-guard invalid waitingFor usage in DockerContainer

### PR DESCRIPTION
**Pull request by wendigo**
Consecutive `waitingFor` calls will override previously defined wait strategies. The correct way of adding multiple wait strategies is to use `waitingForAll` method.

Probably this PR will blow up in the first run. I'll fix invalid usage after first pass.

### CODE REVIEW

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java [hunk 1]
```diff
@@ -83,6 +83,7 @@ public DockerContainer(String dockerImageName, String logicalName)
     {
         super(dockerImageName);
         this.logicalName = requireNonNull(logicalName, "logicalName is null");
+        this.waitStrategy = null; // require explicitly setting waitStrategy
```
Comment (kokosing) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:86]: Optional? Don't you need to override `waitingForAll` as well?

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/suite/suites/SuiteTest.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import io.prestosql.tests.product.launcher.env.EnvironmentConfig;
+import io.prestosql.tests.product.launcher.env.environment.Singlenode;
+import io.prestosql.tests.product.launcher.suite.Suite;
+import io.prestosql.tests.product.launcher.suite.SuiteTestRun;
+
+import java.util.List;
+
+import static io.prestosql.tests.product.launcher.suite.SuiteTestRun.testOnEnvironment;
+
+public class SuiteTest
```
Comment (kokosing) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/suite/suites/SuiteTest.java:26]: Is this related to the commit?

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java (hunk 1 above)
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:86]: nope. this is defined in parent class. waitingForAll calls waitForAll under the hood

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/suite/suites/SuiteTest.java (hunk 2 above)
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/suite/suites/SuiteTest.java:26]: Oops. Removed.

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java (hunk 1 above)
Comment (findepi) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:86]: i am concerned here that you remove a wait strategy for all containers except the ones we have this configured explicitly. is this a step in a right direction? Perhaps. Is it a safe step?

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java [hunk 3]
```diff
@@ -298,6 +299,13 @@ public DockerContainer waitingForAll(WaitStrategy... strategies)
         return this;
     }
 
+    @Override
+    public DockerContainer waitingFor(WaitStrategy waitStrategy)
+    {
+        checkState(this.waitStrategy == null, "waitStrategy is already defined");
```
Comment (findepi) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:305]: i am concerned this will prevent us from reconfiguring a wait strategy along when reconfiguring the container. For example, there is Presto container with some Presto wait strategy using CLI and we now modify the container securing the frontend and we need to amend the wait strategy.

Also, we allow overriding of pretty anything (files, settings, etc), why not wait strategy?

What is the concern you're guarding against here?

Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:305]: I've seen (and corrected) invocations where there was multiple, consecutive waitingFor calls replacing each other (instead waitingForAll should be used). Other solution to this is to merge wait strategies on consecutive calls.

Comment (findepi) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java]: Thanks for explanation. so we can add `waitingFor` override as you did (with a check) and `overrideWaitingFor` to skip the check?
we can (should?) delay adding the new method until needed though

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java [hunk 4]
```diff
@@ -288,14 +288,26 @@ public void copyLogsToHostPath(Path hostPath)
...
+    @Override
+    public DockerContainer waitingFor(WaitStrategy waitStrategy)
+    {
+        if (this.waitStrategy instanceof WaitAllStrategy) {
+            // add to existing wait all strategy
+            return super.waitingFor(((WaitAllStrategy) this.waitStrategy).withStrategy(waitStrategy));
+        }
+        else {
+            // merge existing strategies
+            return super.waitingFor(waitingForAll(this.waitStrategy, waitStrategy));
```
Comment (findepi) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:300]: This in turn disallows replacing the previous wait for.

Again https://github.com/prestosql/presto/pull/6094#discussion_r530668992

> i am concerned this will prevent us from reconfiguring a wait strategy along when reconfiguring the container. For example, there is Presto container with some Presto wait strategy using CLI and we now modify the container securing the frontend and we need to amend the wait strategy.

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java (hunk 3 above)
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java]: I approached that differently. PTAL now

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java [hunk 5]
```diff
@@ -288,14 +289,28 @@ public void copyLogsToHostPath(Path hostPath)
         }
     }
 
-    public DockerContainer waitingForAll(WaitStrategy... strategies)
+    @Override
+    public DockerContainer waitingFor(WaitStrategy waitStrategy)
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:293]: I think the semantics is cumbersome. 
`waitFor` sometimes overrides sometimes adds. 
`watiForAll` always overrides.

I would prefere that override strategy would be explicit in the call. Either in method name, or as an explicit extra parameter.
